### PR TITLE
Clean Duplicates Command

### PR DESCRIPTION
**Pull request by tcitworld**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documentation | needed
| Translation   | no
| Fixed tickets | #2884
| License       | MIT

### CODE REVIEW

File: src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+use Doctrine\ORM\NoResultException;
+use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Wallabag\CoreBundle\Entity\Entry;
+use Wallabag\UserBundle\Entity\User;
+
+class ExportCommand extends ContainerAwareCommand
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php]: Don't forget to change the class name.

Comment (tcitworld) [inline src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php]: Busted.

File: src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+{
+	/** @var OutputInterface */
+	protected $output;
+
+	protected function configure()
+	{
+		$this
+			->setName('wallabag:cleanduplicates')
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php]: I'll prefer `clean-duplicates`, more readable.

File: src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		$this->output = $output;
+
+		$username = $input->getArgument('username');
+
+		if ($username) {
+			try {
+				$user = $this->getUser($username);
+				$this->cleanDuplicates($user);
+			} catch (NoResultException $e) {
+				$output->writeln(sprintf('<error>User "%s" not found.</error>', $input->getArgument('username')));
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php]: You can re-user `$username` instead of the `$input` var

File: src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+				return 1;
+			}
+		} else {
+			$users = $this->getDoctrine()->getRepository('WallabagUserBundle:User')->findAll();
+
+			$output->write(sprintf('Cleaning through %d user accounts', count($users)));
+
+			foreach ($users as $user) {
+				$this->cleanDuplicates($user);
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php]: Maybe output which user you are working one. It'll be easier to debug and check the progress of the command.

File: src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	/**
+	 * @param User $user
+	 */
+	private function cleanDuplicates(User $user)
+	{
+
+		$em = $this->getContainer()->get('doctrine.orm.entity_manager');
+
+		$duplicatesCount = 0;
+		$entries = $user->getEntries();
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Command/CleanDuplicatesCommand.php]: Don't you have performance issue with that? Because I guess it'll retrieve the whole entry and associated relation.
Maybe you should better use a custom repository method to retrieve only the entry id and url.
Then, when you need to remove and entry, you can do it using only its id.

**Comment by tcitworld** [comment]
How can I make more elaborate tests ? (loading duplicates)

**Comment by tcitworld** [comment]
Before adding documentation here I would like to have a *server admin* section in the docs. May we migrate docs, btw nicosomb.
P.S : Gitbook is [great](http://docs.framasoft.org/).

**Comment by nicosomb** [comment]
I gave up my documentation rewrite ... 
I also saw Gitbook. And you're right, it looks great.

**Comment by tcitworld** [comment]
@j0k3r ?
> How can I make more elaborate tests ? (loading duplicates)

File: tests/Wallabag/CoreBundle/Command/CleanDuplicatesCommandTest.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+namespace Tests\Wallabag\CoreBundle\Command;
+
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Component\Console\Tester\CommandTester;
+use Wallabag\CoreBundle\Command\CleanDuplicatesCommand;
+use Tests\Wallabag\CoreBundle\WallabagCoreTestCase;
+
+class CleanDuplicatesCommandTest extends WallabagCoreTestCase
+{
+    public function testRunTagAllCommandForAll()
```
Comment (j0k3r) [inline tests/Wallabag/CoreBundle/Command/CleanDuplicatesCommandTest.php]: `TagAll` ? 🙂

File: src/Wallabag/CoreBundle/Repository/EntryRepository.php [hunk 7]
```diff
@@ -392,4 +392,23 @@ public function getAllEntriesIdAndUrl($userId)
...
+     * Find all entries by url and owner.
+     *
+     * @param $url
+     * @param $userId
+     *
+     * @return array
+     */
+    public function findAllByUrlAndUserId($url, $userId)
+    {
+        $res = $this->createQueryBuilder('e')
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Repository/EntryRepository.php]: You can `return $this->createQueryBuilder('e')`

File: tests/Wallabag/CoreBundle/Command/CleanDuplicatesCommandTest.php [hunk 8]
```diff
@@ -6,10 +6,11 @@
 use Symfony\Component\Console\Tester\CommandTester;
 use Wallabag\CoreBundle\Command\CleanDuplicatesCommand;
 use Tests\Wallabag\CoreBundle\WallabagCoreTestCase;
+use Wallabag\CoreBundle\Entity\Entry;
 
 class CleanDuplicatesCommandTest extends WallabagCoreTestCase
 {
-    public function testRunTagAllCommandForAll()
+    public function testTagAll()
```
Comment (j0k3r) [inline tests/Wallabag/CoreBundle/Command/CleanDuplicatesCommandTest.php]: Nope, it's not the command you are looking for
